src/hf.f90

```diff
@@ -0,0 +1,318 @@
+module hf
+   use const
+
+   implicit none
+
+   type diis_t
+      ! Contains matrices for use in the DIIS procedure
+      logical :: use_diis = .true.
+      integer :: n_errmat = 6
+      integer :: n_active = 0
+      integer :: iter = 0
+      real(p), allocatable :: e(:,:,:)
+      real(p), allocatable :: F(:,:,:)
+      real(p), allocatable :: B(:,:)
+      real(p), allocatable :: c(:)
+      real(p), allocatable :: rhs(:)
+   end type diis_t
+
+   contains
+      subroutine do_rhf(sys, int_store)
+         use error_handling, only: error
+         use system, only: system_t, state_t
+         use integrals, only: int_store_t
+         use linalg, only: mat_t
+
+         type(int_store_t), intent(in) :: int_store
+         type(system_t), intent(inout) :: sys
+
+         type(mat_t) :: ovlp, fockmat
+         real(p) :: ortmat(sys%nbasis,sys%nbasis)
+         real(p) :: density(sys%nbasis,sys%nbasis)
+
+         type(state_t) :: st
+         integer :: iter, iunit, i
+         logical :: conv = .false.
+         type(diis_t) :: diis
+
+         iunit = 6
+         write(iunit, '(1X, 23("-"))')
+         write(iunit, '(1X, A)') 'Restricted Hartree-Fock'
+         write(iunit, '(1X, 23("-"))')
+
+         allocate(st%density_old(sys%nbasis,sys%nbasis), source=0.0_p)
+
+         ! ############################################
+         ! Build the symmetric orthogonalisation matrix
+         ! ############################################ 
+         ! X = S^-1/2 = U s^{-1/2} U^{\dagger} (Szabo and Ostlund eq. 3.167)
+         ! Initialise the overlap matrix stored in int_store_t into a mat_t object
+         call init_mat_t(int_store%ovlp, ovlp)
+         ! We're borrowing this from the SCF diagonalisation type, where the init_mat_t puts ovlp into ao but 
+         ! it's ao_ort that gets passed into BLAS for diagonalisation
+         ovlp%ao_ort = ovlp%ao
+         call diagonalise(ovlp)
+         ortmat(:,:) = 0.0_p
+         ! ovlp%W holds the eigenvalues s_i of S
+         associate(diag=>ovlp%W)
+            do i = 1, size(ortmat, dim=1)
+               ortmat(i,i) = 1/sqrt(diag(i))

```

Comment:
For large bases this might end up blowing up owing to near linear dependence.  Might be worth at least checking the value of diag(i) isn't too small, and throwing an error if it is.